source/grid/grid_tools.cc

```diff
@@ -4352,284 +4351,102 @@ namespace GridTools
-                        break;
-                      }
-                  }
-
-                if (my_pair.first != box_cell[cell_candidate_idx].second)
-                  {
-                    // The cell was not among the candidates returned by the
-                    // tree
-                    cell_candidate_idx = box_cell.size();
-                    box_cell.push_back(
-                      std::make_pair(my_pair.first->bounding_box(),
-                                     my_pair.first));
-                  }
+                missing_points_out.emplace_back(id);
+                // Don't look anymore for this point
+                found_points[id] = true;
               }
           }
+      };
 
+    // If a hint cell was given, use it
+    if (cell_hint.state() == IteratorState::valid)
+      check_all_points_within_box(
+        std::make_pair(mapping.get_bounding_box(cell_hint), cell_hint));
 
-        // Assuming the point is more likely to be in the last
-        // used cell
-        if (my_pair.first == cells_out.back())
-          {
-            // Found in the last cell: adding the data
-            qpoints_out.back().emplace_back(my_pair.second);
-            maps_out.back().emplace_back(p);
-          }
-        else
-          {
-            // Check if it is in another cell already found
-            typename std::vector<typename Triangulation<dim, spacedim>::
-                                   active_cell_iterator>::iterator cells_it =
-              std::find(cells_out.begin(), cells_out.end() - 1, my_pair.first);
-
-            if (cells_it == cells_out.end() - 1)
-              {
-                // Cell not found: adding a new cell
-                cells_out.emplace_back(my_pair.first);
-                qpoints_out.emplace_back(1, my_pair.second);
-                maps_out.emplace_back(1, p);
-              }
-            else
-              {
-                // Cell found: just adding the point index and qpoint to the
-                // list
-                unsigned int current_cell = cells_it - cells_out.begin();
-                qpoints_out[current_cell].emplace_back(my_pair.second);
-                maps_out[current_cell].emplace_back(p);
-              }
-          }
-      }
+    // Now loop over all points that have not been found yet
+    for (unsigned int i = 0; i < np; ++i)
+      if (found_points[i] == false)
+        {
+          // Get the closest cell to this point
+          const auto leaf = b_tree.qbegin(bgi::nearest(points[i], 1));
+          // Now checks all points that fall within this box
+          if (leaf != b_tree.qend())
+            check_all_points_within_box(*leaf);
+          else
+            {
+              // We should not get here. Throw an error.
+              Assert(false,
+                     ExcInternalError(
+                       "I cannot find the closest cell to a point."
+                       "Something is rotten inside."));
```

Comment:
I did not know that you can pass a different than the default text to `ExcInternalError`. I'd try to stick with the default -- you're not providing much more information here